### PR TITLE
feat: BID/ASK stuff

### PR DESCRIPTION
**Pull request by 3Hren**
This PR adds a bunch of code associated with new BID/ASK protocol.

Added:
- Ability to set miner's properties.
- Ability to slice miner into slots.
- Ability to propose deal (incomplete due to missing integration with ETH and market).
- CLI to manage all of this.

Note, that this is still not quite usable due to lack of ETH and market integration as like as deal approving.

Content of `properties.yaml`:
```yaml
foo: bar
cycles: 42
```

How to set properties:
```bash
⇒  ./sonmcli miner --addr="localhost:10001" properties set 3762ffd3-6b06-4572-a0bb-ee19315ceef7 properties.yaml
OK
```

How to get miner properties with miner ID specified:
```bash
⇒  ./sonmcli miner --addr="localhost:10001" properties get 3762ffd3-6b06-4572-a0bb-ee19315ceef7
{"3762ffd3-6b06-4572-a0bb-ee19315ceef7":{"cycles":"42","foo":"bar"}}
```

Without arguments all miners with their properties are shown:
```bash
⇒  ./sonmcli miner --addr="localhost:10001" properties get
{"3762ffd3-6b06-4572-a0bb-ee19315ceef7":{"cycles":"42","foo":"bar"}}
```

Slots API:
```bash
⇒  ./sonmcli miner --addr="localhost:10001" slot add 888616cb-b7ea-4023-8147-694cf1f5895c slot.yaml
OK

⇒  ./sonmcli miner --addr="localhost:10001" slot show
{"888616cb-b7ea-4023-8147-694cf1f5895c":[{"startTime":{"seconds":1512086400},"endTime":{"seconds":1512172800},"buyerRating":100,"supplierRating":42,"resources":{"cpuCores":1,"ramBytes":100000000,"storage":1000000000,"props":{"cycles":"42","foo":"bar"}}}]}

⇒  ./sonmcli miner --addr="localhost:10001" slot show 888616cb-b7ea-4023-8147-694cf1f5895c
{"888616cb-b7ea-4023-8147-694cf1f5895c":[{"startTime":{"seconds":1512086400},"endTime":{"seconds":1512172800},"buyerRating":100,"supplierRating":42,"resources":{"cpuCores":1,"ramBytes":100000000,"storage":1000000000,"props":{"cycles":"42","foo":"bar"}}}]}

⇒  ./sonmcli miner --addr="localhost:10001" slot show OMG
Error: rpc error: code = NotFound desc = miner not found

⇒  ./sonmcli miner --addr="localhost:10001" slot add 888616cb-b7ea-4023-8147-694cf1f5895c slot.yaml
Error: rpc error: code = Unknown desc = specified slot already exists
```

An example of `slot.yaml` content:

```yaml
duration:
  since: 2017-12-01T00:00:00Z
  until: 2017-12-03T00:00:00Z

rating:
  buyer: 100
  supplier: 42

resources:
  cpu_cores: 1
  ram_bytes: 100000000
  gpu_count: 0
  storage: 1000000000

  network:
    in: 0
    out: 0
    type: NO_NETWORK

  properties:
    foo: bar
    cycles: 42
```

### CODE REVIEW

File: insonmnia/hub/server.go [hunk 1]
```diff
@@ -513,7 +515,51 @@ func (h *Hub) ProposeDeal(ctx context.Context, request *pb.DealRequest) (*pb.Dea
...
+
+	if len(h.miners) == 0 {
+		return nil, ErrMinerNotFound
+	}
+
+	rg := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	id := 0
+	var result *MinerCtx = nil
+	for _, miner := range h.miners {
```
Comment (sshaman1101) [inline insonmnia/hub/server.go]: replace `_` with `id`?

**Comment by sshaman1101** [comment]
Please move `cmd/cli/config/slot.go` to `cmd/cli/task_config` because `task_config` directory contains descriptions of loadable yaml files.

File: proto/insonmnia.proto [hunk 2]
```diff
@@ -114,6 +114,12 @@ message TaskResourceRequirements {
     int64 nanoCPUs = 4;
 }
 
+message Empty {}
+
+message UniqueID {
```
Comment (sshaman1101) [inline proto/insonmnia.proto]: Unused message?

File: insonmnia/hub/server.go (hunk 1 above)
Comment (3Hren) [reply insonmnia/hub/server.go]: We need filtered id counter, not total.

File: proto/insonmnia.proto (hunk 2 above)
Comment (3Hren) [reply proto/insonmnia.proto]: Yep.

**Comment by 3Hren** [comment]
Ok.

**Pull request merged**